pints/_mcmc/__init__.py

```diff
@@ -1102,3 +1102,76 @@ def mcmc_sample(log_pdf, chains, x0, sigma0=None, method=None):
     """
     return MCMCController(    # pragma: no cover
         log_pdf, chains, x0, sigma0, method=method).run()
+
+
+def initialise_finite(log_pdf, chains, random_sampler=None, max_n_tries=50,
+                      parallel=False, n_workers=None):
+    """
+    Draws parameter values from a given sampling distribution until either
+    finite values for each of ``chains`` have been generated or the number of
+    attempts exceeds ``num_tries``.
+
+    If log_pdf is of :class:`LogPosterior`, then the
+    `log_pdf.log_prior().sample` method is used for initialisation, although
+    this is overruled by `random_sampler` if it is supplied.
+
+    Parameters
+    ----------
+    log_pdf : pints.LogPDF
+        A :class:`LogPDF` function that evaluates points in the parameter
+        space.
+    chains : int
+        The number of initial values to generate.
+    random_sampler : stochastic function
+        A function that when called returns draws from a probability
+        distribution of same dimensionality as `log_pdf`. The only argument
+        to this function should be an integer specifying the number of draws.
+    max_n_tries : int
+        Number of attempts to find a finite initial value for each of
+        ``chains``.
+    parallel : Boolean
+        Whether to evaluate log_pdf in parallel (defaults to False).
+    n_workers : int
+        Number of workers on which to run parallel evaluation.
+    """
+    if random_sampler is not None and not callable(random_sampler):
+        raise ValueError("random_sampler must be a callable function.")
+
+    if random_sampler is None:
+        if isinstance(log_pdf, pints.LogPosterior):
+            random_sampler = log_pdf.log_prior().sample
+        else:
+            raise ValueError("If log_pdf not of class pints.LogPosterior " +
+                             "then random_sampler must be supplied.")
+
+    if chains < 1:
+        raise ValueError("chains must be 1 or more.")
+
+    if parallel:
+        n_workers = min(pints.ParallelEvaluator.cpu_count(), chains)
+        evaluator = pints.ParallelEvaluator(log_pdf, n_workers=n_workers)
+    else:
+        evaluator = pints.SequentialEvaluator(log_pdf)
+
+    max_n_tries_chains = max_n_tries * chains
+    initialised_finite = False
+    x0 = []
+    n_tries = 0
+    while not initialised_finite and n_tries < max_n_tries_chains:
+        xs = random_sampler(chains)
+        fxs = evaluator.evaluate(xs)
+        xs_iterator = iter(xs)
```

Comment:
I like these iterators!

But some alternatives anyway:
```
for i, x in enumerate(xs):
  fx = fxs[i]
  or fx = next(fxs_iterator)
```
Using more python tricks:
```
for x, fx in zip(xs, fxs):
  print('wow!')
```
I didn't know about that one until a few weeks ago

But probably we want a matrixy version?
```
x0 = np.zeros(n_chains)
while/for ...
  xs = ..
  fxs = evaluate(xs)
  ok = np.is_finite(fxs)
  x0[ok] = xs[ok]
  chains_remaining -= np.sum(ok)
```

Reply:
Ha, thanks for the suggestions: I'm learning new things! I personally find mine a bit more readable (for a non-Python person) but happy to be argued otherwise...